fft_filter.cpp

```diff
@@ -0,0 +1,232 @@
+			processFFT(out);
+		}
+		else
+		{
+			out.push_back(*reading);
+		}
+	}
+	readings->clear();
+}
+
+
+/**
+ * Add a reading to the buffer of data required for an FFT
+ *
+ * @param asset	The asset to buffer
+ */
+void FFTFilter::addFFTAsset(Reading *reading)
+{
+	vector<Datapoint *> datapoints = reading->getReadingData();
+	for (int i = 0; i < datapoints.size(); i++)
+	{
+		string name = datapoints[i]->getName();
+		DatapointValue dpv = datapoints[i]->getData();
+		double value;
+	       	if (dpv.getType() == DatapointValue::T_FLOAT)
+			value = dpv.toDouble();
+		else if (dpv.getType() == DatapointValue::T_INTEGER)
+			value = dpv.toInt();
+		else
+			continue;
+		map<string, vector<double>* >::iterator it = m_buffer.find(name);
+		if (it != m_buffer.end())
+		{
+			it->second->push_back(value);
+		}
+		else
+		{
+			vector<double> *vec = new vector<double>;
+			vec->push_back(value);
+			m_buffer.insert(pair<string, vector<double>* >(name, vec));
+		}
+	}
+}
+
+/**
+ * Process FFT buffer. If we have the right number of samples for an FFT
+ * run against the particular data point then run an FFT against that
+ * data set.
+ *
+ * @param out		The output array for FFT data
+ */
+void FFTFilter::processFFT(vector<Reading *>& out)
+{
+	for (map<string, vector<double>* >::iterator it = m_buffer.begin();
+					it != m_buffer.end(); it++)
+	{
+		if (it->second->size() == m_samples)
+		{
+			runFFT(out, it->first, it->second);
+			it->second->clear();
+		}
+	}
+}
+
+/**
+ * Perform the actual FFT conversion on the sample data. Create the frequency bands
+ * data and insert an asset in the output array with the frequency band amplitudes
+ * and the peak frequency.
+ *
+ * @param out		Vector of readings to be sent onwards from the filter
+ * @param dpName	The name of the datapoint the FFT is beign executed on
+ * @param values	The vector of samples on which we will run the FFT
+ */
+void FFTFilter::runFFT(vector<Reading *>& out, const string& dpName, vector<double> *values)
+{
+float 	*data = new float[values->size() * 2];
+
+	/*
+	 * Turn the sampled data points into complex numbers in an
+	 * array.
+	 */
+	for (int i = 0; i < values->size(); i++)
+	{
+		data[i * 2] = (*values)[i];
+		data[(i * 2) + 1] = 0.0;
+	}
+
+	/*
+	 * Execute the FFT algorithm across the sample
+	 */
+	FFT(data, values->size(), 1);
+
+
+	/*
+	 * Process the results of the FFT algorithm, divide the
+	 * contents of that array into a number of buckets that
+	 * represent a range of frequencies. Average the readings
+	 * across that band to create an amplitude for the band.
+	 *
+	 * Also determine the peak frequency in the data.
+	 */
+	vector<Datapoint *> datapoints;
+	int	aveSamples = values->size() / (2 * m_bands);
+	int	band = 0;
+	double	sum = 0.0;
+	int	cnt = 0;
+	double	peak = 0.0;
+	long	peakf = 0;
+	/*
+	 * Now take the absolute values of the complex numbers to
+	 * get the amplitudes for each frequency
+	 */
+	FILE *fp = fopen("/tmp/fft.csv", "a");
```

Comment:
Is the file usage strict necessary? Could it be a memory based operation?
If possible we can add an option for this.

Reply:
It was there for debug and is now not there any more.